File: golang/leetcode/typescript/longest_valid_parentheses.ts
let LongestValidParentheses = function() {
    this.longestValidParentheses = function(s) {
        let stack = [];
        let max = 0;
        for (let i = 0; i < s.length; i++) {
        if (s[i] == "(" || s[i] == "[" || s[i] == "{") {
            stack.push(i);
        } else {
            if (stack.length == 0) {
            continue;
            }
            let index = stack.pop();
            if (s[index] == "(" && s[i] == ")" || s[index] == "[" && s[i] == "]" || s[index] == "{" && s[i] == "}") {
            max = Math.max(max, i - index + 1);
            } else {
            stack.push(index);
            }
        }
        }
        return max;
    };
};

// test
let longestValidParentheses = new LongestValidParentheses();
console.log(longestValidParentheses.longestValidParentheses("(()"));
console.log(longestValidParentheses.longestValidParentheses("()(()"));
console.log(longestValidParentheses.longestValidParentheses("()(()"));
console.log(longestValidParentheses.longestValidParentheses("()(()"));
console.log(longestValidParentheses.longestValidParentheses("()(()"));
    

File: golang/leetcode/ruby/longest_valid_parentheses.rb
def longestValidParentheses(s)
  return 0 if s.length < 2
  stack = []
  max = 0
  s.each_char.with_index do |c, i|
    if c == '('
      stack.push(i)
    else
      if stack.length > 0
        stack.pop
        max = [max, i - stack.last].max
      else
        stack.push(i)
      end
    end
  end
  max
end

# test
puts longestValidParentheses('(()')
puts longestValidParentheses(')()())')
puts longestValidParentheses(')()())')